### DISCUSSION TITLE
How to set random seed for data transforms?

### DISCUSSION THREAD

**Discussion by Jingnan-Jia** [Project-MONAI/MONAI#2785] [Q&A] [answered]
It seems that `set_determinism` can set a seed for model training, which I think is not enough to exectly reproduce the results of many projects because we still need to set seeds for data transforms during training. 

I do not know how to set the seed of data transforms. Because each transform will be applied to all training images and the seeds would be different between different images (and for images in different epochs).

**Comment by Nic-Ma** [comment, marked as answer]
Hi @Jingnan-Jia ,

The `set_determinism()` API sets random seed for overall program, you can take a look at this tutorial as example:
https://github.com/Project-MONAI/tutorials/blob/master/3d_segmentation/spleen_segmentation_3d.ipynb

Thanks.

**Reply by Jingnan-Jia** [reply]
Got it. Thanks.

**Reply by Jingnan-Jia** [reply]
@Nic-Ma Hi, sorry to bother you again. I want to know if The `set_determinism()` API sets random seed for overall program even if I used my own customed transforms? My own customed transforms are as follows:

```
class RandNoised:
    def __init__(self, *args, **kargs):
        self.noise = np.random.rand()
    def __call__(self, data: Mapping[Hashable, np.ndarray]) -> Dict[Hashable, np.ndarray]:
        d = dict(data)
        d['image'] = d['image'] + self.noise
        return d
```

I guess your answer would be 'No' because I used `np.random.rand` instead of `self.R.rand` which could be obtained if I inherited `Randomizable` class from `Monai`. 

It is easier for me to revise my own customed transforms, but the transforms from torchIO is difficult to replace all `np.random` to `self.R`. So do you think that if we use transforms from torchIO, it is impossible to use `set_determinism()` API sets random seed for overall program.

**Reply by Nic-Ma** [reply]
Hi @Jingnan-Jia ,

The `set_determinism` API will set random seed for global environment, not `self.R`:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/utils/misc.py#L244
If you want to set seed for a particular transform, you should call `transform.set_random_state()`.

Thanks.
